plutus-benchmark/flat/Codec.hs

```diff
@@ -0,0 +1,71 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PackageImports    #-}
+
+module Codec ( Codec(..)
+             , flatCodec
+             , cborCodec
+             , withZlib
+             , withPureZlib
+             , codecs
+             ) where
+
+import           Control.DeepSeq                     (NFData, force)
+import qualified Data.ByteString                     as BS
+import qualified Data.ByteString.Lazy                as LBS
+import           Data.Either                         (fromRight)
+import           Data.Text                           (Text)
+
+import           Language.PlutusCore                 (Name (..), Unique (..))
+import           Language.PlutusCore.Universe
+import           Language.UntypedPlutusCore
+import           Language.UntypedPlutusCore.DeBruijn
+
+import           "pure-zlib" Codec.Compression.Zlib  as PureZlib
+import           "zlib" Codec.Compression.Zlib       as Zlib
+import           Codec.Serialise                     (deserialise, serialise)
+import           Flat                                (DecodeException (..), flat, unflat)
+
+type Tm = Term Name DefaultUni ()
+
+data Codec a = Codec
+  { serialize   :: a -> BS.ByteString
+  , deserialize :: BS.ByteString -> a
+  }
+
+fromDecoded :: Show error => Either error a -> a
+fromDecoded (Left err) = error $ show err
+fromDecoded (Right  v) = v
+
+flatCodec :: Codec (Tm)
+flatCodec = Codec
+  { serialize   = force . flat
+  , deserialize = force . fromDecoded . unflat
+  }
+
+cborCodec :: Codec (Tm)
+cborCodec = Codec
+  { serialize   = force . LBS.toStrict . serialise
+  , deserialize = force . deserialise . LBS.fromStrict
+  }
+
+withZlib :: NFData a => Codec a -> Codec a
+withZlib codec = Codec
+  { serialize = force . LBS.toStrict . Zlib.compress . LBS.fromStrict . serialize codec
```

Comment:
I guess there's a risk of us doing conversions to-and-from lazy bytestrings a lot here? I'd be surprised if GHC was able to eliminate those. But I think that's probably okay for the benchmarks.

If it does matter, maybe we could use lazy bytestrings throughout the codecs, and do any `force`s we need to do at the very end where we use them?

Reply:
I am not sure we care that much about CPU performance when working with Zlib (do we?).

Reply:
Aren't we tracking time as well as space? I think it probably matters a little.

Reply:
There is one way of measuring it, by running the benchmarks. I'll do that and get back to you.

Reply:
After running the benchmarks there is no discernable difference between using and not using `force` and/or `performMajorGC`, so I removed them.

Reply:
I think criterion forces garbage collection just before every execution of the thing you're benchmarking, so it's probably safe to remove it.